### PR TITLE
Exec order of sub commands bug fixes

### PR DESCRIPTION
**Pull request by Asa-Kosto-QTM**

### CODE REVIEW

**Review by neal-erickson — DISMISSED** [review]

File: pytket/phir/phirgen_parallel.py [hunk 1]
```diff
@@ -31,25 +31,26 @@
 
 
 def exec_order_preserved_helper(
-    ordered_dict: OrderedDict["UnitID", int], group_number: int, qubit_last_group: int
+    group_exec_order: list[int], group_number: int, qubit_last_group: int
 ) -> bool:
     """A helper to determine whether order is preserved when adding qubits to groups."""
     # determine whether the current group number is later in execution
     # than the last group in which a qubit was used
```
Comment (qartik) [inline pytket/phir/phirgen_parallel.py:38]: Is this comment describing the role of this helper? If so, it should be part of the docstring.

File: pytket/phir/phirgen_parallel.py [hunk 2]
```diff
@@ -31,25 +31,26 @@
 
 
 def exec_order_preserved_helper(
```
Comment (qartik) [inline pytket/phir/phirgen_parallel.py:33]:
```suggestion
def exec_order_preserved(
```

File: pytket/phir/phirgen_parallel.py [hunk 3]
```diff
@@ -99,7 +104,7 @@ def process_sub_commands(
                 # is that group later in execution than the
                 # most recent group for an op on that qubit?
                 order_preserved = exec_order_preserved_helper(
```
Comment (qartik) [inline pytket/phir/phirgen_parallel.py:106]:
```suggestion
                order_preserved = exec_order_preserved(
```

File: pytket/phir/phirgen_parallel.py [hunk 4]
```diff
@@ -61,25 +62,29 @@ def process_sub_commands(
...
     )
-    num_scs_per_qubit = {}
+    num_scs_per_qubit: dict["UnitID", int] = {}
+    group_exec_order: list[int] = []
 
     for qubit in sub_commands:
         num_scs_per_qubit[qubit] = len(sub_commands[qubit])
         # set every qubit's group id to be -4
         # prevents KeyError in check for group number
         # will get set to a valid group number the first time the qubit is used
```
Comment (qartik) [inline pytket/phir/phirgen_parallel.py:79]: Is this comment still applicable?

**Review by qartik — CHANGES_REQUESTED** [review]
Some questions to clean up comments and code.

File: pytket/phir/phirgen_parallel.py [hunk 5]
```diff
@@ -31,25 +31,26 @@
...
-    return order_preserved
+    # if the group that the qubit is eligible for is later in the exec_order list than
+    # the last group in which it was used, it can be parallelized
+    return group_exec_order.index(group_number) > group_exec_order.index(
+        qubit_last_group
+    )
 
 
-def process_sub_commands(
+def process_sub_commands(  # noqa: PLR0914
```
Comment (qartik) [inline pytket/phir/phirgen_parallel.py:53]:
```suggestion
def process_sub_commands(
```
I think can eliminate the variable `order_preserved` below.

File: pytket/phir/phirgen_parallel.py [hunk 6]
```diff
@@ -61,25 +62,29 @@ def process_sub_commands(
     # and different gate types don't go in the same group
     # RZ gates go in (mod 3)=0 groups, R1XY gates go in (mod 3)=1 groups,
     # and all other gates will go in (mod 3)=2 groups
-    rz_group_number = -3  # will be set to 0 when first RZ gate is assigned (-3 + 3 = 0)
-    r1xy_group_number = (
-        -2  # will be set to 1 when first R1XY gate is assigned (-2 + 3 = 1)
+    rz_group_number: int = -3  # set to 0 when first RZ gate is assigned (-3 + 3 = 0)
+    r1xy_group_number: int = (
```
Comment (qartik) [inline pytket/phir/phirgen_parallel.py:66]: These `int` annotations are unnecessary, same for line 69.

**Comment by qartik** [comment]
Here's a patch containing most of my suggestions, but not all: 
[parallel.patch](https://github.com/CQCL/pytket-phir/files/14195398/parallel.patch)

File: pytket/phir/phirgen_parallel.py (hunk 4 above)
Comment (Asa-Kosto-QTM) [reply pytket/phir/phirgen_parallel.py]: lines 81-84

File: pytket/phir/phirgen_parallel.py [hunk 7]
```diff
@@ -30,23 +30,24 @@
...
-    ordered_dict: OrderedDict["UnitID", int], group_number: int, qubit_last_group: int
+def exec_order_preserved(
+    group_exec_order: list[int], group_number: int, qubit_last_group: int
 ) -> bool:
-    """A helper to determine whether order is preserved when adding qubits to groups."""
-    # determine whether the current group number is later in execution
-    # than the last group in which a qubit was used
+    """A helper to determine whether current group number a qubit is eligible
+    or is later in execution than the last group in which a qubit was used.
+    """  # noqa: D205
```
Comment (qartik) [inline pytket/phir/phirgen_parallel.py:38]:
```suggestion
    """Determine whether order is preserved when adding qubits to groups."""
```
Let's keep it simple as you have a similar comment on lines 46--47 below.

File: pytket/phir/phirgen_parallel.py [hunk 8]
```diff
@@ -30,23 +30,24 @@
...
+    if (group_number not in group_exec_order) or (
+        qubit_last_group not in group_exec_order
+    ):
+        return True
     group_eligible = group_number > qubit_last_group
     if not group_eligible:
         return False
```
Comment (qartik) [inline pytket/phir/phirgen_parallel.py:45]: Would it make sense to simplify these lines as follows:
```suggestion
    if group_number <= qubit_last_group:
        return False
```

I find negative conditions pretty hard to read.

**Review by qartik — APPROVED** [review]
LGTM

**Review by neal-erickson — APPROVED** [review]

**Pull request merged**